plugins/wme-sitebuilder/wme-sitebuilder/Cards/SiteVisibility.php

```diff
@@ -0,0 +1,179 @@
+			'restrictAccess' => $this->getData()->get( self::FIELD_RESTRICT_ACCESS, false ),
+			'password'       => $this->getData()->get( self::FIELD_PASSWORD, '' ),
+		];
+
+		return $details;
+	}
+
+	/**
+	 * Save data.
+	 *
+	 * Hide my sites from search engines, and whatever else.
+	 * Restrict access to visitors with the password.
+	 *  - Password
+	 *
+	 * @return array
+	 */
+	public function save() {
+		$fields = [
+			self::FIELD_HIDE_FROM_SEARCH_ENGINES,
+			self::FIELD_RESTRICT_ACCESS,
+			self::FIELD_PASSWORD,
+		];
+
+		foreach ( $fields as $field ) {
+			$method_name = sprintf( 'set%s', ucfirst( $field ) );
+
+			if ( ! method_exists( $this, $method_name ) ) {
+				// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_trigger_error
+				trigger_error( sprintf( 'Method <code>%s</code> to save <code>%s</code> field is not defined.', esc_html( $method_name ), esc_html( $field ) ), E_USER_WARNING );
+
+				continue;
+			}
+
+			$callable = [ $this, $method_name ];
+
+			if ( ! is_callable( $callable ) ) {
+				// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_trigger_error
+				trigger_error( sprintf( 'Method <code>%s</code> to save <code>%s</code> field is defined but not callable.', esc_html( $method_name ), esc_html( $field ) ), E_USER_WARNING );
+
+				continue;
+			}
+
+			// phpcs:ignore WordPress.Security.NonceVerification.Missing
+			if ( ! array_key_exists( $field, $_REQUEST ) ) {
+				// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_trigger_error
+				trigger_error( sprintf( 'Field <code>%s</code> is absent in $_REQUEST global.', esc_html( $field ) ), E_USER_WARNING );
+
+				continue;
+			}
+
+			// phpcs:ignore WordPress.Security.NonceVerification.Missing
+			call_user_func( $callable, $_REQUEST[ $field ] );
+		}
+
+		// Save the data.
+		$this->getData()->save();
+
+		if ( ! empty( $this->errors ) ) {
+			wp_send_json_error( [
+				'errors' => $this->errors,
+			] );
+		}
+
+		wp_send_json_success();
+	}
+
+	/**
+	 * Set hide from search engines.
+	 *
+	 * @param string $value
+	 */
+	public function setHideFromSearchEngines( $value ) {
+		$current = $this->getData()->get( 'hideFromSearchEngines' );
+
+		if ( $current === $value ) {
+			return;
+		}
+
+		$this->getData()->set( 'hideFromSearchEngines', $value );
+
+		// Update the WordPress option.
+		update_option( 'blog_public', $value == 1 ? 0 : 1 );
+
+		// Error handling.
+		if ( $value != ( get_option( 'blog_public' ) == 1 ? 0 : 1 ) ) {
+			$this->errors[] = __( 'Failed to update the WordPress option <code>blog_public</code>.', 'wme-sitebuilder' );
+		}
+	}
+
+	/**
+	 * Set restrict access.
+	 *
+	 * @param string $value
+	 */
+	public function setRestrictAccess( $value ) {
+		$current = $this->getData()->get( 'restrictAccess' );
+
+		if ( $current === $value ) {
+			return;
+		}
+
+		$this->getData()->set( 'restrictAccess', $value );
+
+		do_action( 'wme_sitebuilder_site_visibility_set_restrict_access', $value, $this );
```

Comment:
Unless you can think of a specific reason to pass it through, let's keep `$this` out of the action to avoid any potential memory leaks with SiteVisibility Card being stored into the hooks arrays and held until the end of the call cycle.